lib/index.js

```diff
@@ -11,76 +9,70 @@ var html =
 '<body>' +
   '<script type="text/javascript">' +
     '(function(window, document) {' +
-      'console.log("WMRM START");' +  // TODO: Remove this
       'var redirectURI = "{TARGET_ORIGIN}";' +
       'var webMessageRequest = {WEB_MESSAGE_REQUEST};' +
       'var authorizationResponse = {' +
         'type: "authorization_response",' +
         'response: {RESPONSE}' +
       '};' +
-'console.log(window.opener); console.log(window.parent);' + // TODO REmove
-'console.log(window.opener != window);' + // TODO REmove
-//      'var mainWin = (window.opener != window) ? window.opener : window.parent;' +
-// NOTE: this is a modification from the draft.  ensure correctness
-      'var mainWin = (window.opener) ? window.opener : window.parent;' +
+
+   // NOTE: this is a modification from the draft in case window.opener is null.
+   // 'var mainWin = (window.opener != window) ? window.opener : window.parent;' +
+      'var mainWin = (window.opener && (window.opener != window)) ? window.opener : window.parent;' +
+
       'if (webMessageRequest["web_message_uri"] && webMessageRequest["web_message_target"]) {' +
-'console.log("relay mode...");' +
         'window.addEventListener("message", function(evt) {' +
-'console.log("GOT EVT AZ");' +
-'console.log(evt);' +
-'console.log(evt.origin);' +
-'console.log(redirectURI);' +
-          'if (evt.origin != redirectURI)' +
+          // NOTE: This is a modification of the draft, and is a weak check. It does not protect
+          // against replay attacks if the attacker shares the same hostname and port as the victim
+          // This is because evt.origin does not contain the full origin, but only the root.
+          'if (evt.origin != redirectURI && evt.origin != webMessageRequest["web_message_root"])' +
```

Comment:
This is the imperfect solution that probably requires us to limit the use of the response mode.